apps/arweave/src/ar_pool.erl

```diff
@@ -0,0 +1,764 @@
+	%% We are going to validate the VDF data later anyways; here we simply want to
+	%% make sure the upper bound is positive so that the recall byte calculation
+	%% does not fail as it takes a remainder of the division by partition upper bound.
+	case PartitionUpperBound > 0 of
+		true ->
+			process_partial_solution_poa(Solution, Ref, H0, H1);
+		_ ->
+			#partial_solution_response{ status = <<"rejected_bad_poa">> }
+	end.
+
+process_partial_solution_poa(Solution, Ref, H0, H1) ->
+	#mining_solution{
+		partition_number = PartitionNumber,
+		partition_upper_bound = PartitionUpperBound,
+		nonce = Nonce,
+		recall_byte1 = RecallByte1,
+		poa1 = PoA1,
+		mining_address = MiningAddress,
+		solution_hash = SolutionH,
+		recall_byte2 = RecallByte2,
+		poa2 = PoA2
+	} = Solution,
+	{RecallRange1Start, RecallRange2Start} = ar_block:get_recall_range(H0,
+			PartitionNumber, PartitionUpperBound),
+	ComputedRecallByte1 = RecallRange1Start + Nonce * ?DATA_CHUNK_SIZE,
+	{BlockStart1, BlockEnd1, TXRoot1} = ar_block_index:get_block_bounds(ComputedRecallByte1),
+	BlockSize1 = BlockEnd1 - BlockStart1,
+	case RecallByte1 == ComputedRecallByte1 andalso
+			ar_poa:validate({BlockStart1, RecallByte1, TXRoot1, BlockSize1, PoA1,
+					{spora_2_6, MiningAddress}, not_set}) of
+		error ->
+			?LOG_ERROR([{event, pool_failed_to_validate_proof_of_access}]),
+			#partial_solution_response{ status = <<"rejected_bad_poa">> };
+		false ->
+			#partial_solution_response{ status = <<"rejected_bad_poa">> };
+		{true, ChunkID} when H1 == SolutionH ->
+			PoACache = {{BlockStart1, RecallByte1, TXRoot1, BlockSize1,
+					{spora_2_6, MiningAddress}}, ChunkID},
+			PoA2Cache = undefined,
+			process_partial_solution_difficulty(Solution, Ref, PoACache, PoA2Cache);
+		{true, ChunkID} ->
+			ComputedRecallByte2 = RecallRange2Start + Nonce * ?DATA_CHUNK_SIZE,
+			{BlockStart2, BlockEnd2, TXRoot2} = ar_block_index:get_block_bounds(
+					ComputedRecallByte2),
+			BlockSize2 = BlockEnd2 - BlockStart2,
+			case RecallByte2 == ComputedRecallByte2 andalso
+					ar_poa:validate({BlockStart2, RecallByte2, TXRoot2, BlockSize2,
+									PoA2, {spora_2_6, MiningAddress}, not_set}) of
+				error ->
+					?LOG_ERROR([{event, pool_failed_to_validate_proof_of_access}]),
+					#partial_solution_response{ status = <<"rejected_bad_poa">> };
+				false ->
+					#partial_solution_response{ status = <<"rejected_bad_poa">> };
+				{true, Chunk2ID} ->
+					PoA2Cache = {{BlockStart2, RecallByte2, TXRoot2, BlockSize2,
+							{spora_2_6, MiningAddress}}, Chunk2ID},
+					PoACache = {{BlockStart1, RecallByte1, TXRoot1, BlockSize1,
+							{spora_2_6, MiningAddress}}, ChunkID},
+					process_partial_solution_difficulty(Solution, Ref, PoACache, PoA2Cache)
+			end
+	end.
+
+process_partial_solution_difficulty(Solution, Ref, PoACache, PoA2Cache) ->
+	#mining_solution{ solution_hash = SolutionH } = Solution,
+	Diff = ar_node:get_current_diff(),
+	case binary:decode_unsigned(SolutionH, big) > Diff of
```

Comment:
This checks the network difficulty, right? Where do we check the partial difficulty? (did I miss the check above somewhere?)